### ISSUE TITLE
Implement preview window

### ISSUE THREAD

**Issue by madskristensen** [madskristensen/WebCompiler#5]
Web Essentials 2013 had a preview window that would show the compiled output side-by-side with the source file. However, the implementation in WE wasn't good enough, so new logic has to be invented for this

**Comment by svenvNL** [comment]
Visual Studio has (or can have) typescript, less, linting. My suggestion would be to compile when those linters/error handlers/whatever say it is possible. Waiting to the point where a user saves takes to long and make the documents out of sync.

I am probably thinking to easily again, but this would in my opinion be the ideal situation as far as I can think of right now.

**Comment by rjgotten** [comment]
> My suggestion would be to compile when those linters/error handlers/whatever say it is possible. Waiting to the point where a user saves takes to long and make the documents out of sync.

The Less compiler can take several seconds to compile a large file with lots of imports. Forcing recompilation on each and every change is a baaaaad idea.

Moreover the Less syntax checker in Visual Studio (even in VS 2015) is woefully out-of-date in syntax support:
- It doesn't understand detached ruleset syntax.
- It doesn't understand guard condition syntax on CSS selectors (only on mixins).
- It doesn't understand the `@plugin` keyword.
- Mixin parameters that appear in the guard condition on the same mixin are erroneously treated as undefined variables.
- It does not understand scope bubbling of variables (which is how 'return values' are implemented).
- It does not understand scope unlocking of mixins.
- etc. etc. etc.

Making recompilation conditional to that syntax checker? No thank you! It would stop the preview window cold in its tracks for any Less file that exhibits any moderate level of complexity.

**Comment by ihaveworms** [comment]
Generating a preview whenever possible (I guess this would mean on any change that doesn't cause a compile error) sounds a little messy. 

There could be a timer implemented so when a user modifies the file (types characters) a timer is fired. Each keystroke resets the timer and after X amount of seconds, it is compiled and a preview generated. This feature could be turned off and a preview shown only on save. A save would always compile.

**Comment by rjgotten** [comment]
> Each keystroke resets the timer and after X amount of seconds, it is compiled and a preview generated. 

You don't even need to monitor key strokes. There's a way to grab an 'application became idle' event from a dispatcher available to VS extensions. I think I saw that in the WE 2013 codebase somewhere before.

If there's also a companion 'application no longer idle' event, you could use that to debounce the recompilation to take place if VS remains idle for X milliseconds.

**Comment by stanislavromanov** [comment]
I don't think that re-compiling every keystroke is a good idea since it might slow down the IDE a lot. However re-compiling on every save is a must. And side-by-side window was one of the best features of WebEssentials. +1

**Comment by rjgotten** [comment]
> And side-by-side window was one of the best features of WebEssentials. +1

It was really the killer feature. Everything else is easily achieved using build scripts (and later: task runners). A side by side preview compilation? Not so much...

**Comment by kelps** [comment]
What if the preview simply displayed the already compiled files, when they exist? This would make it a lot better diagnose problems then what we have now. Then later we could improve it to show live preview compilation on change.

**Comment by justindyer** [comment]
Is this still being worked on or considered?

**Comment by stanislavromanov** [comment]
@justindyer Doubt it, 2 years have passed already. I try to not rely on extension anymore.
